### DISCUSSION TITLE
Using codeql for Cross Compiled environment

### DISCUSSION THREAD

**Discussion by axlev** [github/codeql#14239] [Q&A] [answered]
Hi,
Is cross compilation supported for C/C++?
Trying to run the database creation either with direct or indirect build ends up with:

/bin/sh: /lib/i386-linux-gnu/libc.so.6: version `GLIBC_2.17' not found (required by codeql/tools/linux64/lib32trace.so)

The libc installed both for 32 and 64 bit architecture on the host machine:
ldd --version
ldd (Ubuntu GLIBC 2.35-0ubuntu3.3) 2.35

Is that possible that libc for the target machine not found ? The target is i386 / x64 as well, but the rootfs is different. 

Thanks!
Alex

**Comment by aibaars** [comment, marked as answer]
Cross compilation should be possible. The error message normally means that your libc is too old, so I'm a bit surprised to see this error message when you seem to have version 2.35 installed. Could you double check the stuff that is installed in the rootfs? Also make sure that the target is actually glibc based and not based on something like muslc .

**Comment by axlev** [comment]
The target is based on glibc, the version is actually higher than 2.17:

root@localhost:~# ldd --version
ldd (Debian GLIBC 2.19-18+deb8u10) 2.19

@aibaars Is the error message relates to host or target glibc version ?
Thanks!

**Reply by aibaars** [reply]
That's a lot older than 2.35 but still newer than the required 2.17 . Do you have multiple versions of glibc installed? Have you inspected ` /lib/i386-linux-gnu/libc.so.6` (both host and target) with `ldd`? 

Could you also have a look in the `database/log/build-tracer.log` file? Hopefully, there is some information in there to help diagnose what is going wrong.

**Reply by aibaars** [reply]
> @aibaars Is the error message relates to host or target glibc version ?

It depends, it's probably a mix. CodeQL uses LD_PRELOAD to inject its `libARCHtrace.so` libraries into all binaries that run as part of the build to intercept the compiler invocations (`gcc` and friends). Likely the main build script is running on the host, while the actual `gcc` calls are likely the binaries from the target.

**Comment by axlev** [comment]
We have a single repo from which few different images are created 2 of them seem to have ver 2.13 - indeed older than 2.17. However there is also image with libc 2.19.
Q1: are libc older than 2.17 not supported ? Many embedded systems are still running it as far as I know.
Q2: Is it possible to skip the files with unsupported libc and create database for supported only ? I think it is possible to do it indirectly by using start-tracing and end-tracing, is that the right way to address it ?
Thanks, @aibaars !

**Reply by aibaars** [reply]
> Q1: are libc older than 2.17 not supported ? Many embedded systems are still running it as far as I know.

That's right. The CodeQL tracer requires glibc >= 2.17 to run.  I guess things could work fine if you run the build on a modern OS but compile against the header files of the embedded system. 

> Q2: Is it possible to skip the files with unsupported libc and create database for supported only ? I think it is possible to do it indirectly by using start-tracing and end-tracing, is that the right way to address it ?

I think the easiest would be to only run the build commands for the supported targets.
